### PR TITLE
Freesurfer longitudinal:  REVIEW ONLY

### PR DESCRIPTION
**Pull request by hjmjohnson**
DO NOT INCLUDE THIS!  It is only for review.  This branch should be deleted and  updated later.

### CODE REVIEW

File: nipype/interfaces/freesurfer/utils.py [hunk 1]
```diff
@@ -1276,4 +1276,1108 @@ def _gen_outfilename(self):
...
+
+    _cmd = "mri_fill"
+    input_spec = MRIFillInputSpec
+    output_spec = MRIFillOutputSpec
+    
+    def _list_outputs(self):
+        outputs = self._outputs().get()
+        outputs["out_file"] = os.path.abspath(self.inputs.out_file)
+        return outputs
+
```
Comment (chrisfilo) [inline nipype/interfaces/freesurfer/utils.py]: Codacy Issue found: [class already defined line 1165](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/825382778/issues/source?bid=2767207&fileBranchId=2768049#l1469)

File: nipype/interfaces/freesurfer/utils.py [hunk 2]
```diff
@@ -1276,4 +1276,1108 @@ def _gen_outfilename(self):
...
+class ExtractMainComponentInputSpec(FSTraitedSpec):
+    in_file = File(argstr="%s", position=-2, mandatory=True, exists=True,
+                   desc="Input file for ExtractMainComponent")
+    out_file = File(argstr="%s", position=-1, mandatory=True, exists=False,
+                   desc="Output file for ExtractMainComponent")
+    
+class ExtractMainComponentOutputSpec(TraitedSpec):
+    out_file = File(exists=False, desc="Output file for ExtractMainComponent")
+
```
Comment (chrisfilo) [inline nipype/interfaces/freesurfer/utils.py]: Codacy Issue found: [class already defined line 1175](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/825382778/issues/source?bid=2767207&fileBranchId=2768049#l1478)

File: nipype/interfaces/freesurfer/base.py [hunk 3]
```diff
@@ -151,11 +195,28 @@ def _gen_fname(self, basename, fname=None, cwd=None, suffix='_fs',
...
+    """ Support for Freesurfer script commands with log inputs.terminal_output """
+    _terminal_output = 'file'
+    _always_run = False
+
+    def __init__(self, **inputs):
+        super(FSScriptCommand, self).__init__(**inputs)
+        self.set_default_terminal_output(self._terminal_output)
+
+    def _list_outputs(self):
+        outputs = self.output_spec().get()
```
Comment (chrisfilo) [inline nipype/interfaces/freesurfer/base.py]: Codacy Issue found: [self.output_spec is not callable](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/825382731/issues/source?bid=2767207&fileBranchId=2768049#l209)

File: nipype/interfaces/freesurfer/tests/test_utils.py [hunk 4]
```diff
@@ -241,3 +237,33 @@ def test_surfshots():
...
+
+    # Now test a basic command line
+    adder.inputs.xfm_file = filelist[0]
+    adder.inputs.in_file = filelist[1]
+    outfile = os.path.join(testdir, "b_output.nii")
+    yield assert_equal, adder.cmdline, "mri_add_xform_to_header a.nii b.nii %s" % outfile
+    # Now test that optional inputs get formatted properly
+    adder.inputs.copy_name = True
+    yield assert_equal, adder.cmdline, "mri_add_xform_to_header -c a.nii b.nii %s" % outfile
+    adder.inputs.verbose
```
Comment (chrisfilo) [inline nipype/interfaces/freesurfer/tests/test_utils.py]: Codacy Issue found: [Statement seems to have no effect](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/825382777/issues/source?bid=2767207&fileBranchId=2768049#l265)

File: nipype/interfaces/freesurfer/utils.py [hunk 5]
```diff
@@ -1276,4 +1276,1108 @@ def _gen_outfilename(self):
...
+        outputs = self._outputs().get()
+        outputs["out_file"] = os.path.abspath(self.inputs.out_file)
+        return outputs
+
+class ExtractMainComponentInputSpec(FSTraitedSpec):
+    in_file = File(argstr="%s", position=-2, mandatory=True, exists=True,
+                   desc="Input file for ExtractMainComponent")
+    out_file = File(argstr="%s", position=-1, mandatory=True, exists=False,
+                   desc="Output file for ExtractMainComponent")
+
```
Comment (chrisfilo) [inline nipype/interfaces/freesurfer/utils.py]: Codacy Issue found: [class already defined line 1172](https://www.codacy.com/app/krzysztof-gorgolewski/nipy_nipype/file/825382778/issues/source?bid=2767207&fileBranchId=2768049#l1475)

File: nipype/interfaces/base.py [hunk 6]
```diff
@@ -1567,7 +1567,7 @@ def _filename_from_source(self, name, chain=None):
             retval = name_template % base
             _, _, ext = split_filename(retval)
             if trait_spec.keep_extension and ext:
-                return retval
+                return retval + ext
```
Comment (chrisfilo) [inline nipype/interfaces/base.py]: I'm afraid this is why the tests fail...

**Comment by hjmjohnson** [comment]
@dgellis90  David:  Please review these comments and ensure that your recent patches update the comments by Chris 
@chrisfilo:  Chris:  Thank you for the comments.  We have been making progress on this, but have not updated this patchset recently.

**Pull request closed without merging**